#+TITLE: Reproducing Analysis of {{{tool}}}
#+AUTHOR: Karl Gemayel
#+Date: May 21 2020 
#+OPTIONS: toc:2 H:3 num:3

#+LATEX_HEADER_EXTRA:  \usepackage{mdframed}
#+LATEX_HEADER_EXTRA: \BeforeBeginEnvironment{minted}{\begin{mdframed}}
#+LATEX_HEADER_EXTRA: \AfterEndEnvironment{minted}{\end{mdframed}}
#+LATEX_HEADER: \usepackage[margin=0.5in]{geometry}

#+MACRO: tool MetaGeneMark2

# #+SUBTITLE: The commands used to set up, reproduce, and graph results from the {{{tool}}} paper


* Introduction
This document serves as a step-by-step instruction manual on how to replicate results from the {{{tool}}} paper. 

* Downloading and installing
** Code
Downloading the code is fairly straightforward using =git=. The latest version can be downloaded from =WEBSITE=. To install, simply run 
#+begin_src bash
source config.sh
source install.sh
#+end_src

The first command loads all environment variables (including paths to data directories, binaries, etc...), and the second command creates an executable from all python files and stores them in =$bin= for easy access. For non-unix users, you can find the python driver files in =$driverpython=.
** External Tools
   {{{tool}}} and {{{tool}}}+ and their analysis rely on a handful of external tools. The following need to be installed:
   - GeneMarkS-2
     - Used for building {{{tool}}}+ predictions, and for analysis
     - Link: http://exon.gatech.edu/GeneMark/license_download.cgi
   - ClustalO:
     - Used for constructing multiple sequence alignments
     - Link: http://www.clustal.org/omega/#Download
   - Prodigal:
     - Used for initial analysis of gene-start prediction status
     - Link: https://github.com/hyattpd/Prodigal

** Data
   TODO

   
* Code and data structure

After installing {{{tool}}}, you will have the following structure:

#+begin_src dot :file dir.pdf :cmdline -Tpdf
  digraph{
    sbsp -> data;
    sbsp -> runs;
    sbsp -> lists;
    sbsp -> code;
    sbsp -> bin

    code -> python;
    code -> bash;


    data -> G1;
    data -> G2;
    data -> "...";
    data -> GN;

    G1 -> "sequence.fasta";
    G1 -> "ncbi.gff";
    G1 -> "verified.gff";

    python -> lib;
    python -> driver;
    }
#+end_src

#+RESULTS:
[[file:dir.pdf]]


The =bin= directory contains all executables related to {{{tool}}}, while the =bin_external= may contain external tools, such as GeneMarkS-2 or Prodigal. 

The =data= directory will contain raw genome files (sequence and annotation labels) downloaded from NCBI. In particular, upon initial download of the code, it should contain the genomic sequences for the genomes with experimentally verified gene-starts.

The =list= directory has files that contain different lists of genomes (for example, those with verified genes, those selected as NCBI query genomes, etc...)

Finally the =runs= directory will contain runs of different tools, such as {{{tool}}}, GeneMarkS-2, or Prodigal (as well as one for NCBI's =PGAP=). These will be placed in a subdirectory per genome, as shown below.

#+begin_src dot :file dir_runs.pdf :cmdline -Tpdf
  digraph {
    gms21 [label="gms2"]
    sbsp1 [label="mgm2"]
    prodigal1 [label="prodigal"]

    gms22 [label="gms2"]
    sbsp2 [label="mgm2"]
    prodigal2 [label="prodigal"]

  
  runs -> G1;
    runs -> G2;
    runs -> "...";
    runs -> GN;

    G1 -> gms21;
    G1 -> sbsp1;
    G1 -> prodigal1;
    G2 -> gms22;
    G2 -> sbsp2;
    G2 -> prodigal2;

  }
#+end_src

#+RESULTS:
[[file:dir_runs.pdf]]

* Setting up
** *Important*: Python environment
   Scripts to build and analyze results rely on a handful of python packages. The recommended way to install them is to use the =conda= package manager, and simply run 
   #+begin_src bash
     conda env create -f install/conda_mgm2.yaml
   #+end_src 

   
   To activate this python environment, run 
   #+begin_src bash
     conda activate mg-starts
   #+end_src 
   This automatically loads the correct python libraries and executables into =$PATH=.



* Experiments
** Building MGM2 start models
   #+begin_src bash
     $bin/build_mgm_models_from_gms2_models_py.sh --
   #+end_src
** Extract NCBI Protein Homology Predictions
   #+begin_src bash
     pf_gil=$lists/sbsp.list

     awk -F "," '{if (NR > 1 && NF) print $1}' $pf_gil  | while read -r gcfid; do
       mkdir -p $runs/$gcfid/ncbi_ph;
       awk -F "\t" '{if ($2 == "Protein Homology") print }' $data/$gcfid/ncbi.gff > $runs/$gcfid/ncbi_ph/prediction.gff;
     done
   #+end_src
** Complete Genomes
*** Verified Starts
    #+begin_src bash
      # run tools on verified starts

      dn_experiment=complete_verified
      cd $tmp
      mkdir -p $dn_experiment
      cd $dn_experiment

      pf_gil=$lists/verified.list
      pf_mgm_mod=$bin_external/gms2/mgm_11.mod
      pf_mgm2_mod=$bin_external/mgm2/mgm2_11.mod
      pf_prl_options=$config/parallelization_pbs_2.conf

      declare -a tools=(mgm mgm2 mprodigal fgs mga gms2 prodigal);

      for t in "${tools[@]}"; do
        $bin/run_tool_on_genome_list_py.sh --pf-gil $pf_gil --type auto --tool $t --pf-mgm-mod $pf_mgm_mod --pf-mgm2-mod $pf_mgm2_mod --pf-parallelization-options $pf_prl_options
      done


      # collect statistics
      pf_stats=$(pwd)/summary_complete_verified.csv
      $bin/stats_per_gene_py.sh --pf-gil $pf_gil --tools ncbi_ph ncbi verified "${tools[@]}" --pf-parallelization-options $pf_prl_options --pf-output $pf_stats 

      # visualize statistics
      mkdir -p figures
      cd figures

      $bin/viz_stats_per_gene_with_reference_py.sh --pf-data $pf_stats --reference verified --tools "${tools[@]}" 

      cd $base
    #+end_src
*** StartLink+ Starts
    #+begin_src bash
      # run tools on verified starts

      dn_experiment=complete_startlink
      cd $tmp
      mkdir -p $dn_experiment
      cd $dn_experiment

      pf_gil=$lists/sbsp.list
      pf_mgm_mod=$bin_external/gms2/mgm_11.mod
      pf_mgm2_mod=$bin_external/mgm2/mgm2_11.mod
      pf_prl_options=$config/parallelization_pbs_2.conf

      declare -a tools=(mgm mgm2 mprodigal fgs mga gms2 prodigal);

      for t in "${tools[@]}"; do
        $bin/run_tool_on_genome_list_py.sh --pf-gil $pf_gil --type auto --tool $t --pf-mgm-mod $pf_mgm_mod --pf-mgm2-mod $pf_mgm2_mod --pf-parallelization-options $pf_prl_options
      done


      # collect statistics
      pf_stats=$(pwd)/summary_complete_startlink.csv
      $bin/stats_per_gene_py.sh --pf-gil $pf_gil --tools sbsp ncbi ncbi_ph "${tools[@]}" --pf-parallelization-options $pf_prl_options --pf-output $pf_stats 

      # visualize statistics
      mkdir -p figures
      cd figures

      $bin/viz_stats_per_gene_with_reference_py.sh --pf-data $pf_stats --reference sbsp ncbi_ph --tools "${tools[@]}" 

      cd $base
    #+end_src
** Genome Fragments
*** Verified Starts
    #+begin_src bash
      dn_experiment=chunks_verified
      cd $tmp
      mkdir -p $dn_experiment
      cd $dn_experiment

      pf_gil=$lists/verified.list
      pf_mgm_mod=$bin_external/gms2/mgm_11.mod
      pf_mgm2_mod=$bin_external/mgm2/mgm2_11.mod
      pf_prl_options=$config/parallelization_pbs_2.conf

      declare -a tools=(mgm mgm2 mprodigal fgs mga gms2 prodigal);

      pf_runs_summary=$(pwd)/runs_summary_chunks_verified.csv
      for t in "${tools[@]}"; do
        $bin/run_tools_on_chunks_py.sh --pf-gil $pf_gil --tools "${tools[@]}" --pf-mgm2-mod $pf_mgm2_mod --pf-mgm-mod $pf_mgm_mod --pd-work $runs --pf-summary $pf_runs_summary --pf-parallelization-options $pf_prl_options
      done


      # collect statistics
      pf_stats=$(pwd)/summary_chunks_verified.csv
      $bin/stats_per_gene_on_chunk_py.sh --pf-summary $pf_runs_summary  --reference-tools verified ncbi ncbi_ph  --parallelization-options $pf_prl_options --pf-output $pf_stats 

      # visualize statistics
      mkdir -p figures
      cd figures

      $bin/viz_stats_per_gene_with_reference_py.sh --pf-data $pf_stats --reference verified --tools "${tools[@]}" 

      cd $base
    #+end_src
*** StartLink Starts
    #+begin_src bash
      dn_experiment=chunks_startlink
      cd $tmp
      mkdir -p $dn_experiment
      cd $dn_experiment

      pf_gil=$lists/sbsp.list
      pf_mgm_mod=$bin_external/gms2/mgm_11.mod
      pf_mgm2_mod=$bin_external/mgm2/mgm2_11.mod
      pf_prl_options=$config/parallelization_pbs_2.conf

      declare -a tools=(mgm mgm2 mprodigal fgs mga gms2 prodigal);

      pf_runs_summary=$(pwd)/runs_summary_chunks_startlink.csv
      for t in "${tools[@]}"; do
        $bin/run_tools_on_chunks_py.sh --pf-gil $pf_gil --tools "${tools[@]}" --pf-mgm2-mod $pf_mgm2_mod --pf-mgm-mod $pf_mgm_mod --pd-work $runs --pf-summary $pf_runs_summary --pf-parallelization-options $pf_prl_options
      done


      # collect statistics
      pf_stats=$(pwd)/summary_chunks_startlink.csv
      $bin/stats_per_gene_on_chunk_py.sh --pf-summary $pf_runs_summary  --reference-tools ncbi sbsp sbsp_plus ncbi_ph  --parallelization-options $pf_prl_options --pf-output $pf_stats 

      # visualize statistics
      mkdir -p figures
      cd figures

      $bin/viz_stats_per_gene_with_reference_py.sh --pf-data $pf_stats --reference sbsp ncbi_ph --tools "${tools[@]}" 

      cd $base
    #+end_src
